src/device_gemv.cc

```diff
@@ -0,0 +1,205 @@
+#include "blas/counter.hh"
+
+namespace blas {
+
+//==============================================================================
+namespace impl {
+
+//------------------------------------------------------------------------------
+/// Mid-level templated wrapper checks and converts arguments,
+/// then calls low-level wrapper.
+/// @ingroup gemv_internal
+///
+template <typename scalar_t>
+void gemv(
+    blas::Layout layout,
+    blas::Op trans,
+    int64_t m, int64_t n,
+    scalar_t alpha,
+    scalar_t const* A, int64_t lda,
+    scalar_t const* x, int64_t incx,
+    scalar_t beta,
+    scalar_t*       y, int64_t incy,
+    blas::Queue& queue )
+{
+#ifndef BLAS_HAVE_DEVICE
+    throw blas::Error( "device BLAS not available", __func__ );
+#else
+    using std::swap;
+
+    // check arguments
+    blas_error_if( layout != Layout::ColMajor &&
+                   layout != Layout::RowMajor );
+    blas_error_if( trans != Op::NoTrans &&
+                   trans != Op::Trans &&
+                   trans != Op::ConjTrans );
+    blas_error_if( m < 0 );
+    blas_error_if( n < 0 );
+
+    if (layout == Layout::ColMajor)
+        blas_error_if( lda < m );
+    else
+        blas_error_if( lda < n );
+
+    blas_error_if( incx == 0 );
+    blas_error_if( incy == 0 );
+
+    #ifdef BLAS_HAVE_PAPI
+        // PAPI instrumentation
+        counter::dev_gemv_type element;
+        memset( &element, 0, sizeof( element ) );
+        element = { trans, m, n };
+        counter::insert( element, counter::Id::dev_gemv );
+
+        double gflops = 1e9 * blas::Gflop< scalar_t >::gemv( m, n );
+        counter::inc_flop_count( (long long int)gflops );
+    #endif
+
+    // convert arguments
+    device_blas_int m_    = to_device_blas_int( m );
+    device_blas_int n_    = to_device_blas_int( n );
+    device_blas_int lda_  = to_device_blas_int( lda );
+    device_blas_int incx_ = to_device_blas_int( incx );
+    device_blas_int incy_ = to_device_blas_int( incy );
+
+    blas::internal_set_device( queue.device() );
+
+    // Deal with layout. RowMajor ConjTrans needs copy of x in x2;
+    // in other cases, x2 == x.
+    scalar_t* x2 = const_cast< scalar_t* >( x );
+    Op trans_ = trans;
+    if (layout == Layout::RowMajor) {
+        if constexpr (is_complex_v<scalar_t>) {
+            if (trans == Op::ConjTrans) {
+                // conjugate alpha, beta, x (in x2), and y (in-place)
+                alpha = conj( alpha );
+                beta  = conj( beta );
+
+                x2 = new scalar_t[ m ];
+                int64_t ix = (incx > 0 ? 0 : (1 - m) * incx);
+                for (int64_t i = 0; i < m; ++i) {
+                    x2[ i ] = conj( x[ ix ] );
+                    ix += incx;
+                }
+                incx_ = 1;
+
+                int64_t iy = (incy > 0 ? 0 : (1 - n) * incy);
+                for (int64_t i = 0; i < n; ++i) {
+                    y[ iy ] = conj( y[ iy ] );
+                    iy += incy;
+                }
```

Comment:
Per our email, we will need a routine to conjugate a vector on the GPU for this code. Cf. LAPACK [lacgv](https://www.netlib.org/lapack/explore-html/d9/d50/group__lacgv.html).